File: Python/problem032.py
#! /usr/bin/env python3
"""
Problem 32 - Project Euler
http://projecteuler.net/index.php?section=problems&id=032
"""
from itertools import permutations


def joint(s: list[str]) -> int:
    return int("".join(s))


if __name__ == "__main__":
    ans_set = {
        joint(p[5:9])
        for p in permutations("123456789")
        for i in (1, 2)
        if joint(p[:i]) * joint(p[i:5]) == joint(p[5:9])
    }
    print(sum(ans_set))

File: Ruby/problem032.rb
#! /usr/bin/env ruby
=begin
Problem 32 - Project Euler
http://projecteuler.net/index.php?section=problems&id=32
=end
def pandigital_product(cs)
    product = cs[5, cs.size].join.to_i
    pp =
        (1..3).map {|i|
            cs[0, i].join.to_i * cs[i, 5-i].join.to_i
        }.select { |m| m == product }
        .to_a
    return pp.empty? ? nil : product
end

puts '123456789'.chars
                .permutation
                .map {|p| pandigital_product(p)}
                .compact
                .uniq
                .sum
